stdlib/header.c

```diff
@@ -13,42 +13,111 @@
 /**************************************************************************/
 
-#define CAML_INTERNALS
+/* The launcher for bytecode executables (if #! is not available) */
+
+/* C11's _Noreturn is deprecated in C23 in favour of attributes */
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
+  #define NORETURN [[noreturn]]
+#else
+  #define NORETURN _Noreturn
+#endif
 
-/* The launcher for bytecode executables (if #! is not working) */
+#ifdef _WIN32
+
+#define STRICT
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+
+#if WINDOWS_UNICODE
+#define CP CP_UTF8
+#else
+#define CP CP_ACP
+#endif
+
+/* mingw-w64 has a limits.h which defines PATH_MAX as an alias for MAX_PATH */
+#if !defined(PATH_MAX)
+#define PATH_MAX MAX_PATH
+#endif
+
+#define SEEK_END FILE_END
+
+#define lseek(h, offset, origin) SetFilePointer((h), (offset), NULL, (origin))
+
+static int read(HANDLE h, LPVOID buffer, DWORD buffer_size)
+{
+  DWORD nread = 0;
+  ReadFile(h, buffer, buffer_size, &nread, NULL);
+  return nread;
+}
+
+static BOOL WINAPI ctrl_handler(DWORD event)
+{
+  if (event == CTRL_C_EVENT || event == CTRL_BREAK_EVENT)
+    return TRUE;                /* pretend we've handled them */
+  else
+    return FALSE;
+}
+
+static void write_error(const wchar_t *wstr, HANDLE hOut)
+{
+  DWORD consoleMode, numwritten, len;
+  char str[MAX_PATH];
+
+  if (GetConsoleMode(hOut, &consoleMode) != 0) {
+    /* The output stream is a Console */
+    WriteConsole(hOut, wstr, lstrlen(wstr), &numwritten, NULL);
+  } else { /* The output stream is redirected */
+    len =
+      WideCharToMultiByte(CP, 0, wstr, lstrlen(wstr), str, sizeof(str),
+                          NULL, NULL);
+    WriteFile(hOut, str, len, &numwritten, NULL);
+  }
+}
+
+NORETURN static void exit_with_error(const wchar_t *wstr1,
+                                     const wchar_t *wstr2,
+                                     const wchar_t *wstr3)
+{
+  HANDLE hOut = GetStdHandle(STD_ERROR_HANDLE);
+  if (wstr1) write_error(wstr1, hOut);
+  if (wstr2) write_error(wstr2, hOut);
+  if (wstr3) write_error(wstr3, hOut);
+  write_error(L"\r\n", hOut);
+  ExitProcess(2);
+}
+
+#else
 
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include "caml/s.h"
-#ifndef _WIN32
 #include <unistd.h>
-#endif
 #include <fcntl.h>
+#include <limits.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#include "caml/mlvalues.h"
-#include "caml/exec.h"
 
-#ifndef MAXPATHLEN
-#define MAXPATHLEN 1024
+/* O_BINARY is defined in Gnulib, but is not POSIX */
+#ifndef O_BINARY
+#define O_BINARY 0
 #endif
 
-#ifndef S_ISREG
-#define S_ISREG(mode) (((mode) & S_IFMT) == S_IFREG)
+/* C11's _Noreturn is deprecated in C23 in favour of attributes */
+#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 202311L
+  #define NORETURN [[noreturn]]
+#else
+  #define NORETURN _Noreturn
 #endif
 
-#ifndef SEEK_END
-#define SEEK_END 2
-#endif
+typedef int HANDLE;
```

Comment:
```suggestion
typedef void* HANDLE;
```

Reply:
I don't understand this suggestion - the type _is_ an `int`? The point here is to be using the same type name for file descriptors between each? (i.e. type `HANDLE` is `int` on Unix and `HANDLE` on Windows - if Posix actually defined a type for file descriptors, I'd have done it the other way around)

Reply:
Thinking some more, in the first version I generally tried to make the code either look like POSIX or, otherwise, look like Windows (i.e. define things that looked like CRT functions where needed)... the core point is so that `read_runtime_path` has a sensible looking type (which is needed for both).

Perhaps it would therefore be better to use a completely fresh name: i.e.

```c
#ifdef _WIN32
typedef HANDLE file_descriptor
#else
typedef int file_descriptor
#endif
/* ... */
static char * read_runtime_path(file_descriptor fd)
/* ... */
```
?

Reply:
Sorry, I misunderstood the code, and thought this typedef was in a Windows code path. I like your suggestion of using a different name than `HANDLE` which I wouldn't expect in a Unix code path. `file_descriptor` or `filedescr` are ok, I think.